### PR TITLE
[MOD-3048] Disallow `enable_memory_snapshot=True` with parameterized classes

### PR DESCRIPTION
**Pull request by mattnappo**
Temporary patch until we fix the root problem on the backend.

### CODE REVIEW

File: modal/app.py [hunk 1]
```diff
@@ -723,6 +723,13 @@ def wrapper(user_cls: CLS_T) -> _Cls:
             ):
                 raise InvalidError("A class must have `enable_memory_snapshot=True` to use `snap=True` on its methods.")
 
+            # Disallow enable_memory_snapshot for parameterized classes
+            constructor = dict(inspect.getmembers(user_cls, inspect.isfunction)).get("__init__")
+            if enable_memory_snapshot and constructor:
+                params = inspect.signature(constructor).parameters
+                if len(params) > 1:
+                    raise InvalidError("Cannot use parameterized classes with `enable_memory_snapshot=True`.")
```
Comment (thundergolfer) [inline modal/app.py:731]: I think it'll be more user friendly if the error message includes the name of the class so the user can more quickly go to the problematic source.

Comment (thundergolfer) [reply modal/app.py:731]: I think this code block should also have a TODO which references a Linear ticket (MOD-123) about supporting this properly.

**Review by thundergolfer — APPROVED** [review]
LGTM after addressing comments.

**Comment by mwaskom** [comment]
I'm not up to speed on the issue here but if the problem is in the backend, is this sometihng we could do server-side so we can (a) roll it out to everyone and (b) remove it when we fix the issue?

**Comment by mattnappo** [comment]
Could something like this be in `FunctionBindParams`?

**Comment by mwaskom** [comment]
> Could something like this be in FunctionBindParams?

Yeah that is what I was thinking at first, you could add something that checks the definition to see if snapshots are enabled.

One potential downside is that I don't think it will fire on `modal deploy`, so people might deploy apps that then break when they handle their first input. Maybe that's still better than the status quo though? Again I'm not up to speed with what the problem is.

**Comment by mattnappo** [comment]
That is better than what we are currently doing, but I'd argue for a wider net here (including deploys) since most use cases for parameterized classes is so that you can do a `.lookup()` on a deployed app.

**Comment by mwaskom** [comment]
> since most use cases for parameterized classes is so that you can do a .lookup() on a deployed app

It should still work in this case, the problem is just that it will let you deploy the app in a broken state. Ideally we would block the deploy itself.

File: modal/app.py [hunk 2]
```diff
@@ -723,6 +723,17 @@ def wrapper(user_cls: CLS_T) -> _Cls:
...
 
+            # Disallow enable_memory_snapshot for parameterized classes
+            # TODO(matt) Temporary fix for MOD-3048
+            constructor = dict(inspect.getmembers(user_cls, inspect.isfunction)).get("__init__")
+            if enable_memory_snapshot and constructor:
+                params = inspect.signature(constructor).parameters
+                if len(params) > 1:
+                    name = user_cls.__name__
+                    raise InvalidError(
+                        f"In class '{name}': Cannot use parameterized classes with `enable_memory_snapshot=True`."
```
Comment (thundergolfer) [inline modal/app.py:734]:
```suggestion
                        f"Cannot use class parameterization in class {name} with `enable_memory_snapshot=True`."
```

**Comment by thundergolfer** [comment]
Good point @mwaskom. I don't like that users can get all the way past `modal deploy` with broken if we just have server-side. But it is a little annoying that when we fix it users will need to upgrade client. 

I think I'm in favor of both client and server?

**Comment by mattnappo** [comment]
Here's the backend change: https://github.com/modal-labs/modal/pull/13974

**Pull request merged**